src/apify/scrapy/middlewares/apify_proxy.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import annotations
+
+from base64 import b64encode
+from typing import TYPE_CHECKING
+from urllib.parse import ParseResult, unquote, urlparse
+
+from scrapy.core.downloader.handlers.http11 import TunnelError
+from scrapy.exceptions import NotConfigured
+from scrapy.utils.python import to_bytes
+
+from apify import Actor, ProxyConfiguration
+
+if TYPE_CHECKING:
+    from scrapy import Request, Spider
+    from scrapy.crawler import Crawler
+
+
+class ApifyHttpProxyMiddleware:
+    """Apify HTTP proxy middleware for Scrapy.
+
+    This middleware enhances request processing by adding a 'proxy' field to the request's meta and an
+    authentication header. It draws inspiration from the Scrapy default `HttpProxyMiddleware`. The proxy URL is
+    sourced from the settings under the `APIFY_PROXY_SETTINGS` key. The value of this key, a dictionary, should be
+    provided by the Actor input. An example of the proxy settings:
+
+    proxy_settings = {'useApifyProxy': true, 'apifyProxyGroups': []}
+    """
+
+    def __init__(
+        self: ApifyHttpProxyMiddleware,
+        proxy_settings: dict,
```

Comment:
Could we use a `TypedDict` here? A plain `dict` is difficult to reason about - is it a mapping? Probably not, what keys does it accept?

Reply:
That would be nice. However, it's just an argument that is passed to the [Actor.create_proxy_configuration](https://github.com/apify/apify-sdk-python/blob/master/src/apify/actor.py#L1282) (and then it is passed further),  where it is just a general dictionary as well. We do not do the processing of the dict here.

If we would want a more specific typing here we should create a typedict (or dataclass) for it and use it across all its occurrences.

Reply:
Or even some kind of a model class that does validation. I understand if it's out of the scope of this PR, and I also understand that you don't usually instantiate this class manually. But still, it would make the code more readable.

Reply:
See https://github.com/apify/apify-sdk-python/issues/64